compiler/rustc_resolve/src/late.rs

```diff
@@ -229,6 +229,11 @@ enum LifetimeRibKind {
     /// This rib declares generic parameters.
     Generics { binder: NodeId, span: Span, kind: LifetimeBinderKind },
 
+    /// This rib declares an item-like without generic params.
+    /// This is used to know where `Fresh` lifetime parameters should be introduced
+    /// in `AnonymousCreateParameter` mode.
+    NoGenerics { binder: NodeId },
```

Comment:
I don't understand how this works.
Apparently `NoGenerics` is treated in the same way as `Generics`, but what actually happens if you add lifetime parameters to, for example, a `static`?
Will they be silently ignored or produce an error?

Reply:
There are 2 concepts here.
The place where we put the lifetime parameter is specified by `AnonymousCreateParameter`.  The list of created generics is saved in `extra_lifetime_params: NodeId -> Vec<LifetimeRes>`.
We also need a parent def-id to create a generic parameter's def-id.  However, the place where we put the lifetime may not have one.  This can happen for instance for `fn` types.  In that case, we use the `NoGenerics` rib to the correct "parent" de-if.  This can be used, for instance, if we have
```rust
static Foo: fn(&u8) = |_| {};
               ^- this lifetime has DefId ~ Foo::'_
```

I too am unhappy of the intertwined concepts.  I'm also a bit worried about possible bugs in the presence of existential impl-trait or other exotic stuff.

Reply:
https://github.com/rust-lang/rust/pull/97720/commits/d1aef5408c5cc56e16b6140193065397a6229c19 looks like a much better solution, if lowering is a place where we already know the closest def id for every possible node id, then it's reasonable to do it there.